File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void solve(vector<int>&arr,int target,vector<int>&output,vector<vector<int>>&ans,int index,int n){
          if(index>=n){
            if(target==0)ans.push_back(output);
            return;
          }
          // include;
          if(target-arr[index]>=0){
            output.push_back(arr[index]);
            solve(arr,target-arr[index],output,ans,index,n);
            output.pop_back();
          }
          //exclude
          solve(arr,target,output,ans,index+1,n);


    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int>output;
        vector<vector<int>>ans;
        int n=candidates.size();
        solve(candidates,target,output,ans,0,n);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
   public void solve(int[] candidates,int target,List<Integer>output,List<List<Integer>>ans,int ind){
        if(ind>=candidates.length){
            if(target==0){
                ans.add(new ArrayList<>(output));
            }
            return;
        }
        if(target-candidates[ind]>=0){
            output.add(candidates[ind]);
            solve(candidates,target-candidates[ind],output,ans,ind);
            output.remove(output.size()-1);
        }
        solve(candidates,target,output,ans,ind+1);
   }
      
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<Integer> output=new ArrayList<>();
        List<List<Integer>>ans=new ArrayList<>();
        solve(candidates,target,output,ans,0);
        return ans;
    }
}